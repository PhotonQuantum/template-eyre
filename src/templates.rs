pub const SIMPLE: &str = r#"{{error}}
{{#if (gt (len sources) 0)}}
Caused by:
{{#if (gt (len sources) 1)}}
{{#each sources}}
{{indent @index this}}
{{/each}}
{{else}}
{{#each sources}}
{{indent this}}
{{/each}}
{{/if}}
{{/if}}"#;

pub const COLORED_SIMPLE: &str = r#"{{style "red" error}}
{{#if (gt (len sources) 0)}}
{{style "black.bright" "Caused by:"}}
{{#if (gt (len sources) 1)}}
{{#each sources}}
{{indent @index (style "yellow" this)}}
{{/each}}
{{else}}
{{#each sources}}
{{indent (style "yellow" this)}}
{{/each}}
{{/if}}
{{/if}}"#;
