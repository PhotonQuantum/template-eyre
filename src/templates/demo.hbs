Oh no, this program crashed!

{{style "red" error}}
{{*set multi=(gt (len sources) 1)}}
{{#each sources}}
    {{#if @first}}
        {{~style "black.bright" "Caused by:"}}
    {{/if}}
    {{~indent (_if @root.multi @index null) (style "yellow" this)}}
{{/each}}

{{style "cyan" "Please report this issue to ..."}}