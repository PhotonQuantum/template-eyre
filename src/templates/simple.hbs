{{error}}
{{*set multi=(gt (len sources) 1)}}
{{#each sources}}
    {{#if @first~}}
        Caused by:
    {{/if}}
    {{~indent (_if @root.multi @index null) this}}
{{/each}}