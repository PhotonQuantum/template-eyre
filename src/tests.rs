use std::error::Error;
use std::sync::{Mutex, MutexGuard};

use eyre::{eyre, EyreHandler, Report};
use once_cell::sync::Lazy;

use crate::templates::{COLORED_SIMPLE, SIMPLE};
use crate::Hook;

const DEMO: &str = r#"Oh no, this program crashed!

{{style "red" error}}
Caused by:
{{#each sources}}
{{indent @index (style "yellow" this)}}
{{/each}}

{{style "cyan" "Please report this issue to ..."}}"#;

macro_rules! report {
    [$($e:literal),+] => {
        _report(&[$($e),+])
    }
}

#[test]
fn simple() {
    must_report("simple", SIMPLE);
    must_report("colored_simple", COLORED_SIMPLE);
    must_report("demo", DEMO);
}

macro_rules! assert_snapshot {
    ($name: expr, $value: expr) => {
        insta::assert_snapshot!($name, format!("{:?}", $value));
    };
}

fn must_report(name: &str, template: &str) {
    let _guard = hack_install(Hook::new(template, "").unwrap());
    console::set_colors_enabled(true);

    let report = report!["Unable to talk to daemon"];
    println!("{:?}", report);
    assert_snapshot!(format!("{}_1", name), report);

    let report = report![
        "Unable to talk to daemon",
        "Connection refused (os error 61)"
    ];
    println!("{:?}", report);
    assert_snapshot!(format!("{}_2", name), report);

    let report = report![
        "Unable to talk to daemon",
        "Connection refused",
        "os error 61"
    ];
    println!("{:?}", report);
    assert_snapshot!(format!("{}_3", name), report);
}

type ErrorHook =
    Box<dyn Fn(&(dyn Error + 'static)) -> Box<dyn EyreHandler> + Sync + Send + 'static>;

static DYNAMIC_HOOK: Lazy<Mutex<Option<ErrorHook>>> = Lazy::new(|| Mutex::new(None));
static TEST_GUARD: Lazy<Mutex<()>> = Lazy::new(|| Mutex::new(()));

fn hack_install(hook: Hook) -> MutexGuard<'static, ()> {
    let guard = TEST_GUARD.lock().unwrap();
    let mut dynamic_hook = DYNAMIC_HOOK.lock().unwrap();
    if dynamic_hook.is_none() {
        eyre::set_hook(Box::new(|e| {
            (DYNAMIC_HOOK.lock().unwrap().as_ref().unwrap())(e)
        }))
        .unwrap();
    }
    *dynamic_hook = Some(Box::new(move |e| Box::new(hook.make_handler(e))));
    guard
}

fn _report(errors: &'static [&'static str]) -> Report {
    errors
        .iter()
        .rfold::<Option<Report>, _>(None, |x, acc| {
            x.map_or_else(|| Some(eyre!(acc)), |x| Some(x.wrap_err(acc)))
        })
        .unwrap()
}
