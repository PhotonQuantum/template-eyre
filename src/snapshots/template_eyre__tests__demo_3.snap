---
source: src/tests.rs
assertion_line: 72
expression: "format!(\"{:?}\", report)"

---
Oh no, this program crashed!

[31mUnable to talk to daemon[0m
[2mCaused by:[0m
   0: [33mConnection refused[0m
   1: [33mos error 61[0m

[36mPlease report this issue to ...[0m
